.github/workflows/ci.yaml

```diff
@@ -112,12 +148,24 @@ jobs:
           node-version: 20
       - uses: extractions/setup-just@v1
       - uses: actions/cache@v3
-        name: cache
+        name: toolchain cache
+        with:
+          path: |
+            ~/.rustup/toolchains/
+          key: ${{ runner.os }}-toolchain-${{ hashFiles('**/rust-toolchain.toml') }}
+      - uses: actions/cache@v3
+        name: cargo cache
         with:
           path: |
             ~/.cargo/
+          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+      - uses: actions/cache@v3
+        name: build cache
+        with:
+          path: |
             target/
             !target/**/glaredb
+            !target/**/slt
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
```

Comment:
- I have a memory of each individual cache action invocation taking time, and so I remember in the past I combined them. Could you just check that we're not running into that? (It's possible that I'm being superstitious) 
- I think it makes sense to cache `~/.cargo` against a hash of the `Cargo.lock` file; but I'm not sure that makes the same amount of sense to cache the target directory against the same hash. Should we use the branch name so that different builds (are less likely) to overwrite each other?

Reply:
> I have a memory of each individual cache action invocation taking time,

they do take _some_ time, but its pretty negligible. The build cache is the biggest one, and takes like 40 seconds. the other caches add a total of <15 seconds to the build. 

But what they add up front, they save in the slt pipelines. Now since it is precompiled, we can skip all of the caches except the one containing the binary `slt cache`. So it really allows us to parallelize the stages a lot more without getting penalized by pulling down the entire cache every time. 

> but I'm not sure that makes the same amount of sense to cache the target directory against the same hash. Should we use the branch name so that different builds (are less likely) to overwrite each other?

I was thinking about this too, but without the build cache set to the same key as the cargo cache, the initial build wont benefit from incremental compilation. So it ends up being much slower. 

I definitely think we should experiment further with the caches. I'd be pretty happy with a <10 minute pipeline

Reply:
> I was thinking about this too, but without the build cache set to the same key as the cargo cache, the initial build wont benefit from incremental compilation.

I'm sort of just thinking outloud but could we handle `target/**/incremental` with the cargo cache and the rest of the build/target cache with the branch?

Reply:
i'm not sure if we'd want to try to cache certain parts of `target/`, it may break some things, but not entirely sure. We should follow up in a separate issue/pr.